### PR TITLE
Encode leaf index in second child pointer

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

**Comment by dalg24** [comment]
Haven't added tests for `cast()`.
Haven't added a static assertion that the size of `Node *` matches `size_t`.

File: packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp [hunk 1]
```diff
@@ -254,32 +255,68 @@ void TreeConstruction<DeviceType>::sortObjects(
...
+    // value type set to the size type of the memory space.  Cuda memory space
+    // size type is `unsigned int` while it is `size_t` on the host...
+    auto permute = bin_sort.get_permute_vector();
+
+    // The static assertion below is meant to check whether changes occured in
+    // Kokkos_Sort.hpp or in the definition of the size types in which case we
+    // might want to reconsider what to do with the permutation indices.
+    static_assert(
+#ifdef KOKKOS_ENABLE_CUDA
+        std::is_same<typename decltype( permute )::traits::execution_space,
```
Comment (sslattery) [inline packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp]: Are you worried there will be changes to this in the future?

File: packages/Search/src/details/DTK_DetailsUtils.hpp [hunk 2]
```diff
@@ -252,6 +252,39 @@ typename View::non_const_type cloneWithoutInitializingNorCopying( View &v )
...
+          std::is_same<typename View::value_type, Value>::value>::type * = 0 )
+{
+    return src;
+}
+
+template <typename Value, typename View>
+inline Kokkos::View<Value *, typename View::traits::array_layout,
+                    typename View::traits::device_type,
+                    typename View::traits::memory_traits>
+cast( View const &src,
```
Comment (sslattery) [inline packages/Search/src/details/DTK_DetailsUtils.hpp]: is `cast` the best name for this function? If the value types are different you are allocating and doing a copy. Maybe `changeViewDataType` or something similar?

File: packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp [hunk 3]
```diff
@@ -254,32 +255,68 @@ void TreeConstruction<DeviceType>::sortObjects(
...
+                       sizeof( Node * ),
+                   "Encoding leaf index in pointer to child is not safe if the "
+                   "index and pointer types do not have the same size" );
     Kokkos::parallel_for(
         DTK_MARK_REGION( "initialize_leaf_nodes" ),
         Kokkos::RangePolicy<ExecutionSpace>( 0, n ), KOKKOS_LAMBDA( int i ) {
             leaf_nodes( i ).bounding_box = bounding_boxes( indices( i ) );
-            leaf_nodes( i ).children = {nullptr, nullptr};
+            leaf_nodes( i ).children = {
+                nullptr, reinterpret_cast<Node *>( indices( i ) )};
```
Comment (sslattery) [inline packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp]: So the main change here is now that the child pointer is also the leaf index?

Comment (sslattery) [reply packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp]: But the Morton indices are still composed to build the tree and create a permutation vector for sorting, correct?

Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp]: The pointer to children of a leaf node were null, now we use one of them to store the permutation index.

Nothing has changed as far as the Morton order is concerned.

File: packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp (hunk 1 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp]: I just want to know it if something changes.

File: packages/Search/src/details/DTK_DetailsUtils.hpp (hunk 2 above)
Comment (dalg24) [reply packages/Search/src/details/DTK_DetailsUtils.hpp]: I couldn't come up with anything better.  I don't like `changeViewDataType`.  Anyone has a better idea?

Comment (aprokop) [reply packages/Search/src/details/DTK_DetailsUtils.hpp]: `type_cast`?

File: packages/Search/src/DTK_LinearBVH_def.hpp [hunk 4]
```diff
@@ -42,9 +40,9 @@ BoundingVolumeHierarchy<DeviceType>::BoundingVolumeHierarchy(
 
     if ( size() == 1 )
     {
-        iota( _indices );
+        Kokkos::View<size_t *, DeviceType> permutation_indices( "dummy", 1 );
```
Comment (aprokop) [inline packages/Search/src/DTK_LinearBVH_def.hpp]: I would not call this "dummy". You do care that `permutation_indices[0] = 0`.

Comment (dalg24) [reply packages/Search/src/DTK_LinearBVH_def.hpp]: Sure.  I can change it.  It is a temporary though.  And yes, the single element has to be zero.

**Comment by aprokop** [comment]
> What would you do? using SizeType= typename Kokkos::BinSort<Kokkos::View<unsigned int *, DeviceType>, CompType>::size_type? H no!

"Ugly is beautiful" (Arthur Hailey)

> That said, I realize now that BinSort::size_type is aliasing a template parameter that defaults to the memory space size type. I realize now I can specify it to be size_t and then all the casting hell goes away.

Neat!

**Comment by dalg24** [comment]
@aprokop would you please review the updated version?  I will rebase and squash/drop before the final merge.

**Pull request merged**